collab2/foraging/toolkit/proximity.py

```diff
@@ -1,93 +1,209 @@
+    3. For distances beyond the optimal range, proximity decays exponentially, representing a
+       diminishing effect.
+
+    :param distance: A float or ndarray representing the distance(s) at which proximity is evaluated.
+    :param getting_worse: The distance threshold below which the score becomes negative. Defaults to 1.5.
+    :param optimal: The distance where proximity reaches its peak. Defaults to 4.
+    :param proximity_decay: The rate at which proximity decays beyond the optimal range. Defaults to 1.
+
+    :return: A float or ndarray representing the computed proximity value(s) based on the input distance.
+    """
+
+    cond1 = distance <= getting_worse
+    cond2 = (distance > getting_worse) & (
+        distance <= getting_worse + 1.5 * (optimal - getting_worse)
+    )
+
+    result = np.where(
+        cond1,
+        np.sin(np.pi / (2 * getting_worse) * (distance + 3 * getting_worse)),
+        np.where(
+            cond2,
+            np.sin(
+                np.pi / (2 * (optimal - getting_worse)) * (distance - getting_worse)
+            ),
+            np.sin(
+                np.pi
+                / (2 * (optimal - getting_worse))
+                * (1.5 * (optimal - getting_worse))
+            )
+            * np.exp(
+                -proximity_decay
+                * (distance - optimal - 0.5 * (optimal - getting_worse))
+            ),
+        ),
+    )
+
+    return result
+
+
+# TODO this can be further generalized to any point-centered contribution
+# with some decay/scoring function
+def _proximity_predictor_contribution(
+    x_other: int,
+    y_other: int,
+    grid: pd.DataFrame,
+    proximity_function: Callable,
+    **proximity_function_kwargs,
+) -> np.ndarray:
+    """
+    Computes the proximity score contribution of an agent present at `x_other, y_other` to points present in `grid`.
+
+    This function calculates the Euclidean distance from the point `(x_other, y_other)` to each point
+    in the grid and applies the provided proximity function to determine the proximity score for those points
+    as a function of distance.
+
+    :param x_other: The x-coordinate of an agent.
+    :param y_other: The y-coordinate of an agent.
+    :param grid: A pandas DataFrame containing the x and y coordinates of the grid points. The DataFrame
+                 must have columns named 'x' and 'y'.
+    :param proximity_function: A callable function that takes distance as input and computes proximity scores.
+    :param proximity_function_kwargs: Additional keyword arguments to pass to the proximity function.
+
+    :return: A numpy ndarray representing the computed proximity scores for each point in the grid.
+    """
+
+    distance_to_other = np.sqrt((grid["x"] - x_other) ** 2 + (grid["y"] - y_other) ** 2)
+    proximity_score = proximity_function(distance_to_other, **proximity_function_kwargs)
+    return proximity_score
+
+
+def _proximity_predictor(
```

Comment:
pls rename to `_generate_proximity_predictor` :)

Reply:
done